### PR TITLE
Services client based on kv store

### PR DESCRIPTION
**Pull request by cw9**

### CODE REVIEW

**Comment by cw9** [comment]
Moved sd client code out of statsdex 
@xichen2020 @robskillington

File: services/client/services.go [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+	return service, nil
+}
+
+func (s *client) Watch(sid services.ServiceID, opts services.QueryOptions) (xwatch.Watch, error) {
+	if err := validateRequest(sid); err != nil {
+		return nil, err
+	}
+
+	s.opts.InstrumentsOptions().Logger().Infof(
+		"adding a watch for service: %s env: %s zone: %s includeUnhealthy: %v",
```
Comment (xichen2020) [inline services/client/services.go]: Do we really need to log here? This seems debugging-related

File: services/client/services.go [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+
+func newServiceDiscoveryMetrics(m tally.Scope) serviceDiscoveryMetrics {
+	return serviceDiscoveryMetrics{
+		versionGauge:        m.Gauge("placement.verison"),
+		serviceUnmalshalErr: m.Counter("placement.unmarshal.error"),
+	}
+}
+
+func placementKey(env, service string) string {
+	// e.g [production]m3db
```
Comment (xichen2020) [inline services/client/services.go]: Probably should remove this comment

File: services/client/options.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3x/instrument"
+)
+
+// KVGen generates a kv store for a given zone
+type KVGen func(zone string) (kv.Store, error)
+
+// Options are options for the client of Services
+type Options interface {
+	// InitTimeout is the max time to wait on a new service watch for a valid initial value
```
Comment (xichen2020) [inline services/client/options.go]: Might be worthwhile to add a `Validate` here to check if `KVGen` etc has been set

File: services/client/util.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+		return shard.Available
+	} else if s == placementproto.ShardState_Leaving {
+		return shard.Leaving
+	}
+
+	panic("unknown shard state from proto")
+}
+
+func shardStateToProto(s shard.State) placementproto.ShardState {
+	if s == shard.Initializing {
```
Comment (xichen2020) [inline services/client/util.go]: Might be cleaner to use a `switch`

File: services/client/util.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+	} else if s == shard.Available {
+		return placementproto.ShardState_Available
+	} else if s == shard.Leaving {
+		return placementproto.ShardState_Leaving
+	}
+
+	panic("unknown shard state to proto")
+}
+
+type sortableProtoShards []*placementproto.Shard
```
Comment (xichen2020) [inline services/client/util.go]: Rename to `shardByIDAscending`?

File: services/client/services.go (hunk 1 above)
Comment (cw9) [inline services/client/services.go]: This barely logs, most times only log once when a service starts up, seems to be helpful to me

File: services/client/services.go (hunk 2 above)
Comment (cw9) [inline services/client/services.go]: sure can do

File: services/client/options.go (hunk 3 above)
Comment (cw9) [inline services/client/options.go]: Sure, can do

File: services/client/util.go (hunk 4 above)
Comment (cw9) [inline services/client/util.go]: will do

File: services/client/util.go (hunk 5 above)
Comment (cw9) [inline services/client/util.go]: will do

**Comment by xichen2020** [comment]
LGTM

**Pull request merged**